### PR TITLE
access log: identify downstream disconnects in HCM

### PR DESCRIPTION
**Pull request by venilnoronha**
*Description*: This identifies downstream disconnects from within the HTTP Connection Manager and eventually echoes it out via the access log.
*Risk Level*: Med
*Testing*: Updated tests
*Docs Changes*: N/A
*Release Notes*: N/A
See #5119, #5167 and [this comment](https://github.com/envoyproxy/envoy/pull/5167#discussion_r239179331) for further information.

/cc @mattklein123 @htuch

### CODE REVIEW

File: include/envoy/http/codec.h [hunk 1]
```diff
@@ -115,8 +115,10 @@ enum class StreamResetReason {
   RemoteRefusedStreamReset,
   // If the stream was locally reset by a connection pool due to an initial connection failure.
   ConnectionFailure,
-  // If the stream was locally reset due to connection termination.
-  ConnectionTermination,
+  // If the stream was reset due to an upstream connection termination.
+  UpstreamConnectionTermination,
+  // If the stream was reset due to a downstream connection termination.
+  DownstreamConnectionTermination,
```
Comment (htuch) [inline include/envoy/http/codec.h]: Thanks for taking on the more comprehensive change suggested by @mattklein123. I have one design question here; in `codec.h`, we're largely providing an interface for codecs. We instantiate codecs in HCM for downstream and then again in the codec client for the upstream connection. Shouldn't these reset reasons be independent of concepts like upstream/downstreams?

Comment (mattklein123) [reply include/envoy/http/codec.h]: +1, from an initial read through I wonder if we could do this a bit more simply? Like basically just use the logic you originally had in the other PR in the HCM before the log gets written? (Obviously we should keep the stuctured logging updates which were missing from the last PR).

Comment (venilnoronha) [reply include/envoy/http/codec.h]: Addressed it in 6014815.

**Comment by mattklein123** [comment]
@lizan we need to figure out how to make clang-tidy output something so it doesn't time out. Is there any way to do this? I totally hacky thing would be to have a wrapper script that prints something every so often while the real thing runs. @jmarantz already reported this in https://github.com/envoyproxy/envoy/issues/5144

**Pull request merged**